crates/volta-core/src/tool/yarn/mod.rs

```diff
@@ -0,0 +1,77 @@
+use std::fmt::{self, Display};
+
+use super::{debug_already_fetched, info_fetched, info_installed, info_pinned, Tool};
+use crate::error::ErrorDetails;
+use crate::session::Session;
+use crate::style::tool_version;
+use semver::Version;
+use volta_fail::Fallible;
+
+mod fetch;
+mod resolve;
+mod serial;
+
+pub use resolve::resolve;
+
+/// The Tool implementation for fetching and installing Yarn
+#[derive(Debug)]
+pub struct Yarn {
+    pub(super) version: Version,
+}
+
+impl Yarn {
+    pub fn new(version: Version) -> Self {
+        Yarn { version }
+    }
+
+    pub(crate) fn fetch_internal(&self, session: &mut Session) -> Fallible<()> {
```

Comment:
I'd be inclined here to pass in the required `&mut Inventory` and `&Hooks` dependencies *instead* of `Session`: it makes the contract here clearer, and it decreases our coupling to the current design of the `Session` as our bag for everything (in case we want to refactor in the future). In general I prefer to pass functions as minimal and exact a set of what they need as possible, for that reason.

Reply:
See below for some of the reasoning behind passing the `Session` object. In this case, I think there's less of an issue if we wanted to pass the needed objects directly, the only concern I have is then each of the Trait methods that call `fetch_internal` will have to independently resolve the `&mut Inventory` and `&Hooks` objects, which adds some boilerplate to the overall flow, without a significant benefit.